### ISSUE TITLE
Refactor how BaseCompletionModel.delete_cur_item works.

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#545]
After #531 is merged, we should refactor how custom sorting in models work.

Currently, `BaseCompletionModel` raises `NotImplementedError` in `sort()`, and `CompletionFilterModel` catches this and calls its own sort implementation instead.

This means pylint will mark the class as abstract, and show `abstract-method` for every completion class. Currently we explicitely ignore this everywhere, but this is annoying.

Instead, a model should probably have a `custom_sort` bool member which decides whether the model supplies a sort method, or not.

**Comment by krobelus** [comment]
does this solution make sense? https://github.com/krobelus/qutebrowser/commit/996625adafa2db7a2e947d0092f7f5985d894b0d
i am not sure because i did not manage to get the corresponding pylint errors either way (i ran `pylint ../../../ qutebrowser.completion.models` in `scripts/dev/pylint_checkers`)

the same thing should probably apply to custom_filter() in BaseCompletionModel

**Comment by The-Compiler** [comment]
Hmm, yeah, currently the same happens for `custom_filter` and `delete_cur_item`:

```
************* Module qutebrowser.completion.models.urlmodel
W: 31, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
************* Module qutebrowser.completion.models.configmodel
W: 29, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 29, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 43, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 43, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 96, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 96, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
************* Module qutebrowser.completion.models.miscmodels
W: 31, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 31, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 52, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 52, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 92, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W: 92, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W:104, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W:104, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W:116, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W:116, 0: Method 'delete_cur_item' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
W:131, 0: Method 'custom_filter' is abstract in class 'BaseCompletionModel' but is not overridden (abstract-method)
```

Your solution would work, but I think I need to think about what the best solution for this would be first...

**Comment by lahwaacz** [comment]
There is currently no `custom_filter` other than the stub in the abstract class. Is it needed at all or has the perfect condition for refactoring just been prepared?

**Comment by The-Compiler** [comment]
Oh, right - seems like it was used only in cbc4ec6531f648201a8240e917d3929ad7e9ed37 and then removed in aaa523ce7cce9993c52f53567e2c5a55120b4d74. Time to get rid of it I think! Want to do so, @lahwaacz, or should I?

**Comment by The-Compiler** [comment]
@krobelus I don't think that'll work - there's `QAbstractItemModel::sort` which does nothing, so `hasattr(self.srcmodel, 'sort')` will always be true.

**Comment by lahwaacz** [comment]
I'm already doing some modifications for https://github.com/The-Compiler/qutebrowser/issues/1651, so I can take care of this.

**Comment by lahwaacz** [comment]
There isn't any method overriding `BaseCompletionModel.sort` either, can that be deleted too?

**Comment by The-Compiler** [comment]
Right - looks like that got added in 59948a038c66f13ab33fddcd7ec543ac103af961 and removed in fe4f32606d5572c5295574c74512afba28c790c9 - so yeah, kill it :wink:

**Comment by rcorre** [comment]
@lahwaacz I'm writing tests for sortfilter now and was thinking we should kill custom sorting too. Are you planning on merging the commits linked above?

**Comment by The-Compiler** [comment]
@rcorre Those commits are part of #1654, where I'm really not sure if I should merge it as-is as it seems to worsen completion performance...

**Comment by lahwaacz** [comment]
I can always split the trivial commits (55347bde4474be6b85ad770700bab95b36998d9a, f7e7effb4c46eb52581139682ab661b04780005f, b98064dc16ddbe9b4193582d48f7c7dc72557ccb and maybe abb2f6c9d49a4a964dc983b89940e7011a654bf2) into separate PR and then do a rebase...

**Comment by The-Compiler** [comment]
@lahwaacz that'd be great! But I think 55347bde4474be6b85ad770700bab95b36998d9a already has the changed filtering behaviour in it?

**Comment by lahwaacz** [comment]
> But I think 55347bde4474be6b85ad770700bab95b36998d9a already has the changed filtering behaviour in it?

Yeah, it was rather demanding rebase... But I hope that I got it right in #1705. I'll rebase #1654 on master when #1705 is merged.

**Comment by The-Compiler** [comment]
And gone it is!

**Comment by lahwaacz** [comment]
Maybe the [first comment](https://github.com/The-Compiler/qutebrowser/issues/545#issue-60673223) still aplies on [delete_cur_item](https://github.com/The-Compiler/qutebrowser/blob/master/qutebrowser/completion/models/base.py#L106-L108) though?

**Comment by The-Compiler** [comment]
Hmm, that's true. Really not sure what the cleanest way to handle this kind of thing is.

**Comment by The-Compiler** [comment]
Fixed with #2295 - I'm closing it already so I don't forget.
